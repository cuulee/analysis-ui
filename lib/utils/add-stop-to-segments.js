import getSegment from './get-segment'

export async function atStart ({
  coordinates,
  followRoad,
  isStop,
  segments,
  spacing,
  stopId
}) {
  // insert at start
  let stopAtEnd, to, toStopId

  // handle case of no existing stops
  if (segments.length > 0) {
    to = segments[0].geometry.type === 'LineString'
      ? segments[0].geometry.coordinates[0]
      : segments[0].geometry.coordinates
    // if segments[0] is a point, the from and to stop information are identical (see below) so we don't have to worry about
    // detecting that case here.
    stopAtEnd = segments[0].stopAtStart
    toStopId = segments[0].fromStopId
  } else {
    to = null // leave null so a point is created
    // duplicate all the information so it will be picked up when the next stop is created
    stopAtEnd = isStop
    toStopId = stopId
  }

  const newSegment = await getSegment({
    followRoad,
    from: coordinates,
    fromStopId: stopId,
    // can also be a point if only one stop has been created
    segments,
    spacing,
    stopAtEnd,
    stopAtStart: isStop,
    to,
    toStopId
  })

  segments.unshift(newSegment)
  // if there was a segment that was just a point, get rid of it
  if (segments.length === 2 && segments[1].geometry.type === 'Point') {
    segments.pop()
  }

  return segments
}

export async function inMiddle ({
  coordinates,
  followRoad,
  index,
  isStop,
  segments,
  stopId
}) {
  // replacing one segment with two in the middle
  const sourceSegment = segments[index]
  const [segment0, segment1] = await Promise.all([
    getSegment({
      followRoad,
      from: sourceSegment.geometry.coordinates[0],
      fromStopId: sourceSegment.fromStopId,
      segments,
      spacing: sourceSegment.spacing,
      stopAtEnd: isStop,
      stopAtStart: sourceSegment.stopAtStart,
      to: coordinates,
      toStopId: stopId
    }),
    getSegment({
      followRoad,
      from: coordinates,
      fromStopId: stopId,
      segments,
      spacing: sourceSegment.spacing,
      stopAtEnd: sourceSegment.stopAtEnd,
      stopAtStart: isStop,
      to: sourceSegment.geometry.coordinates.slice(-1)[0],
      toStopId: sourceSegment.toStopId
    })
  ])

  // swap out the segments
  segments.splice(index, 1, segment0, segment1)

  return segments
}

export async function atEnd ({
  coordinates,
  followRoad,
  isStop,
  segments,
  spacing,
  stopId
}) {
  // insert at end
  let from, fromStopId, stopAtStart

  const lastSegIdx = segments.length - 1

  // handle creating the first stop. Note that we only support this when adding at the end.
  if (segments.length > 0) {
    from = segments[lastSegIdx].geometry.type === 'LineString'
      ? segments[lastSegIdx].geometry.coordinates.slice(-1)[0]
      : segments[lastSegIdx].geometry.coordinates
    stopAtStart = segments[lastSegIdx].stopAtEnd
    fromStopId = segments[lastSegIdx].toStopId
  } else {
    from = null
    stopAtStart = isStop
    fromStopId = stopId
  }

  const newSegment = await getSegment({
    followRoad,
    from,
    fromStopId,
    // can also be a point if only one stop has been created
    segments,
    spacing,
    stopAtEnd: isStop,
    stopAtStart,
    to: coordinates,
    toStopId: stopId
  })

  segments.push(newSegment)

  // if there was a segment that was just a point, get rid of it
  if (segments.length === 2 && segments[0].geometry.type === 'Point') {
    segments.shift()
  }

  return segments
}
