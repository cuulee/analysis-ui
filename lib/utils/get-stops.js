// @flow
import dbg from 'debug'
import bearing from '@turf/bearing'
import distance from '@turf/distance'
import {point} from '@turf/helpers'

import type {Segment, Stop} from '../types'

/**
 * Flow stops along a route. When autogenerating stops between bona fide stops,
 * how close can an autogenerated stop come to the next non-auto-generated stop,
 * as a fraction of stop spacing? TODO Make configurable?
 */
const MIN_SPACING_PERCENTAGE = 0.25

const debug = dbg('utils:get-stops')

export default function getStops (segments: Segment[]): Stop[] {
  if (segments.length === 0) return []
  if (segments.length === 1 && segments[0].geometry.type === 'Point') {
    const coord = segments[0].geometry.coordinates
    return [
      {
        center: coord,
        stopId: segments[0].fromStopId,
        index: 0,
        lat: coord[1],
        lon: coord[0],
        bearing: false, // this will hide directional icon, which is what we want
        autoCreated: !segments[0].stopAtStart,
        distanceFromStart: 0
      }
    ]
  }

  const stops = []
  const coord = segments[0].geometry.coordinates[0]
  stops.push({
    center: coord,
    stopId: segments[0].fromStopId,
    index: 0,
    lat: coord[1],
    lon: coord[0],
    bearing: bearing(point(coord), point(segments[0].geometry.coordinates[1])),
    autoCreated: !segments[0].stopAtStart,
    distanceFromStart: 0
  })

  // loop over the route, making stops as we go
  let distanceToLastStop = 0
  let distanceToLineSegmentStart = 0

  for (let segIdx = 0; segIdx < segments.length; segIdx++) {
    const segment = segments[segIdx]
    // now loop over line segments within this segment, accumulating distance as we go
    // a single transit segment can have multiple line segments, because we've used a street router between endpoints
    for (let i = 1; i < segment.geometry.coordinates.length; i++) {
      const c0 = segment.geometry.coordinates[i - 1]
      const c1 = segment.geometry.coordinates[i]
      const distanceThisLineSegment =
        distance(point(c0), point(c1), 'kilometers') * 1000

      // segment.spacing = 0 means no automatic stop creation in this segment
      while (
        segment.spacing > 0 &&
        distanceToLastStop + segment.spacing <
          distanceToLineSegmentStart + distanceThisLineSegment
      ) {
        // how far into the segment do we place the stop
        let frac =
          (distanceToLastStop + segment.spacing - distanceToLineSegmentStart) /
          distanceThisLineSegment
        if (frac < 0) frac = 0 // most likely the last segment did not have automatic stop creation

        const pos = [
          c0[0] + (c1[0] - c0[0]) * frac,
          c0[1] + (c1[1] - c0[1]) * frac
        ]

        // can't just add segment.spacing because of converting negative fractions to zero above
        // this can happen when the last segment did not have automatic stop creation, or had a larger spacing
        // TODO in the latter case, we probably want to continue to apply the spacing from the last line segment until we create a new stop?
        distanceToLastStop =
          distanceToLineSegmentStart + frac * distanceThisLineSegment

        stops.push({
          center: pos,
          stopId: null,
          index: segIdx,
          lat: pos[1],
          lon: pos[0],
          autoCreated: true,
          bearing: bearing(point(c0), point(c1)),
          distanceFromStart: distanceToLastStop
        })
      }

      distanceToLineSegmentStart += distanceThisLineSegment
    }

    if (segment.stopAtEnd) {
      const endCoord = segment.geometry.coordinates.slice(-1)[0]
      stops.push({
        center: endCoord,
        stopId: segment.toStopId,
        index: segIdx + 1,
        lat: endCoord[1],
        lon: endCoord[0],
        autoCreated: false,
        distanceFromStart: distanceToLineSegmentStart,
        bearing: segIdx < segments.length - 1
          ? bearing(
            point(segments[segIdx + 1].geometry.coordinates[0]),
            point(segments[segIdx + 1].geometry.coordinates[1])
          )
          : false
      })

      // restart the spacing
      distanceToLastStop = distanceToLineSegmentStart // distanceToLineSegmentStart already set to the start of the next line segment
    }
  }

  // filter out autocreated stops that are very close to bona fide stops
  return stops.filter((stop, stopIndex, stops) => {
    if (!stop.autoCreated) return true

    const spacingThisSegment = segments[stop.index].spacing
    if (spacingThisSegment === 0) {
      // fail harder?
      debug(
        `segment ${stop.index} contains auto-created stops but does not have automatic stop creation enabled?`
      )
      return true
    }

    // last stop
    if (stopIndex === stops.length - 1) return true

    // only need to check if it's too close to the next stop, as it will always be spaced away from the previous
    // stop, since we flow stops forward from the bona fide stops
    const nextStop = stops[stopIndex + 1]
    const delta = nextStop.distanceFromStart - stop.distanceFromStart
    const spacedCorrectlyFromNextStop =
      nextStop.autoCreated ||
      delta / spacingThisSegment >= MIN_SPACING_PERCENTAGE

    return spacedCorrectlyFromNextStop
  })
}
