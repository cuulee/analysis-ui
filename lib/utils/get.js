// @flow
import get from 'lodash/get'

export default function (o: any, p: string, d: any) {
  const value = get(o, p, d)
  return value === null ? d : value
}
