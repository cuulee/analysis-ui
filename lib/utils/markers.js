import bearing from '@turf/bearing'
import distance from '@turf/distance'
import {point} from '@turf/helpers'

const DIRECTION_MARKER_SPACING_METERS = 1500

/**
 * Get directional markers for a GeoJSON LineString
 */
export function getBearingAndCoordinatesAlongLine ({
  coordinates,
  spacingMeters = DIRECTION_MARKER_SPACING_METERS
}) {
  const markers = [
    {
      bearing: bearing(point(coordinates[0]), point(coordinates[1])),
      coordinates: [coordinates[0][0], coordinates[0][1]]
    }
  ]

  for (
    let distanceToLastMarker = 0, segIdx = 1, totalDistance = 0;
    segIdx < coordinates.length;
    segIdx++
  ) {
    const segStart = coordinates[segIdx - 1]
    const segEnd = coordinates[segIdx]
    const distanceThisSegment =
      distance(point(segStart), point(segEnd), 'kilometers') * 1000

    // while not if, may need multiple markers on a single long segment
    while (
      distanceToLastMarker + spacingMeters < totalDistance + distanceThisSegment
    ) {
      // total distance is at start of this segment, figure out how much of the spacing is in this segment vs. last
      const metersIntoThisSegment =
        distanceToLastMarker + spacingMeters - totalDistance
      const frac = metersIntoThisSegment / distanceThisSegment
      const lat = segStart[1] + (segEnd[1] - segStart[1]) * frac
      const lon = segStart[0] + (segEnd[0] - segStart[0]) * frac

      markers.push({
        bearing: bearing(point(segStart), point(segEnd)),
        coordinates: [lon, lat]
      })

      distanceToLastMarker += spacingMeters
    }
    totalDistance += distanceThisSegment
  }

  return markers
}
