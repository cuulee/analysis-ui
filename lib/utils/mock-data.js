// @flow

import multi from '@conveyal/woonerf/store/multi'
import promise from '@conveyal/woonerf/store/promise'
import {middleware as fetchMiddleware} from '@conveyal/woonerf/fetch'
import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import React from 'react'
import {Provider} from 'react-redux'
import configureStore from 'redux-mock-store'
import thunkMiddleware from 'redux-thunk'
import {TextEncoder} from 'text-encoding'

import {responseToSurface} from '../actions/analysis'
import {ADD_TRIP_PATTERN, TRAVEL_TIME_PERCENTILES} from '../constants'

export const makeMockStore = configureStore([
  fetchMiddleware,
  multi,
  promise,
  thunkMiddleware
])

export const mockWithProvider = (children: any, mockData: any = mockStores.init) => {
  const store = makeMockStore(mockData)
  const wrapper = mount(<Provider store={store}>{children}</Provider>)

  return {
    snapshot: () => mountToJson(wrapper),
    store,
    wrapper
  }
}

export const mockModel = {
  accessGroup: 'accessGroupName',
  createdAt: '2017-11-08T12:30:50.133Z',
  createdBy: 'person@gmail.com',
  nonce: '12345',
  updatedAt: '2017-11-08T12:31:19.602Z',
  updatedBy: 'person@gmail.com'
}

export const mockSegmentGeom = {
  type: 'LineString',
  coordinates: [
    [-122.0246, 36.9707],
    [-122.0279, 37.049],
    [-121.9799, 37.2299],
    [-121.9445, 37.324],
    [-121.936, 37.353],
    [-121.924, 37.365]
  ]
}

export const mockSegment = {
  fromStopId: '1',
  geometry: mockSegmentGeom,
  properties: {},
  spacing: 300,
  stopAtEnd: false,
  stopAtStart: false,
  toStopId: null,
  type: 'Feature'
}

export const mockStop1 = {
  autoCreated: false,
  bearing: false,
  center: [36.9707, -122.0246],
  distanceFromStart: 0,
  index: 0,
  lat: 36.9707,
  lon: -122.0246,
  stopId: '1'
}

const mockGtfsStop1 = {
  stop_id: '1',
  stop_lat: 36.9707,
  stop_lon: -122.0246,
  stop_name: 'Santa Cruz'
}

const mockGtfsStop2 = {
  stop_id: '2',
  stop_lat: 37.049,
  stop_lon: -122.0279,
  stop_name: 'Scotts Valley'
}

export const mockGtfsStops = [mockGtfsStop1, mockGtfsStop2]

export const mockPattern = {
  geometry: mockSegmentGeom,
  name: 'Mock Pattern',
  pattern_id: 'mock-pattern-id',
  stops: mockGtfsStops,
  trips: [
    {
      duration: 1234,
      end_time: 13579,
      start_time: 12345,
      trip_headsign: 'To Downtown',
      trip_id: 'abcd',
      trip_short_name: 'The Express'
    }
  ]
}

export const mockRoute = {
  label: 'mock-route-label',
  patterns: [mockPattern],
  route_id: 'route1'
}

export const mockFeed = {
  checksum: 'MOCK',
  id: '1',
  routes: [mockRoute],
  stopsById: {
    '1': mockGtfsStop1,
    '2': mockGtfsStop2
  },
  stops: [mockGtfsStop1, mockGtfsStop2]
}

export const mockMapState = {
  allowExtend: true,
  extendFromEnd: true,
  followRoad: true,
  state: 'add-trip-pattern',
  modificationId: '1234'
}

export const mockTimetable = {
  _id: 'timetable-id',
  name: 'Test timetable',
  segmentSpeeds: [],
  dwellTime: 10,
  monday: true,
  tuesday: true,
  wednesday: true,
  thursday: true,
  friday: true,
  saturday: false,
  sunday: false,
  headwaySecs: 900, // 15 minutes
  startTime: 28800, // 8am
  endTime: 57600, // 4pm
  patternTrips: ['abcd'],
  sourceTrip: 'abcd',
  phaseFromTimetable: '',
  phaseFromStop: '',
  phaseAtStop: '',
  phaseSeconds: 300,
  exactTimes: false,
  modificationId: '1234',
  modificationName: 'Test Modification'
}

export const mockOpportunityDataset = {
  dataSource: 'lodes-data-2014',
  name: 'Workers total',
  key: 'Workers_total'
}

export const mockRegion = {
  ...mockModel,
  bounds: {
    north: 39,
    east: -76,
    south: 38,
    west: -77
  },
  description: 'Region description',
  _id: '1',
  name: 'Mock Region',
  opportunityDatasets: [mockOpportunityDataset]
}

export const mockProject = {
  ...mockModel,
  bundleId: '1',
  _id: '1',
  name: 'Mock Project',
  regionId: mockRegion._id,
  variants: ['Default']
}

export const mockModification = {
  ...mockModel,
  bidirectional: false,
  description: 'Mock description',
  entries: [],
  feed: mockFeed.id,
  _id: '1234',
  name: 'Test Modification',
  routes: [mockRoute.route_id],
  projectId: mockProject._id,
  segments: [mockSegment],
  showOnMap: false,
  timetables: [mockTimetable],
  trips: ['abcd'],
  type: ADD_TRIP_PATTERN,
  variants: [true]
}

export const mockBundle = {
  ...mockModel,
  name: 'Mock Bundle',
  _id: '1',
  centerLat: 38.90124,
  centerLon: -77.015615,
  feeds: [],
  regionId: mockRegion._id,
  totalFeeds: 0
}

export const mockProfileRequest = {
  date: '2016-01-16',
  fromTime: 25200,
  toTime: 32400,
  accessModes: 'WALK',
  directModes: 'WALK',
  egressModes: 'WALK',
  transitModes: 'BUS,TRAM,RAIL,SUBWAY',
  walkSpeed: 1.3888888888888888,
  bikeSpeed: 4.166666666666667,
  reachabilityThreshold: 0,
  bikeSafe: 1,
  bikeSlope: 1,
  bikeTime: 1,
  monteCarloDraws: 200,
  maxRides: 4,
  maxTripDurationMinutes: 60,
  travelTimePercentile: 50
}

export const mockStores = {
  init: {
    analysis: {
      isFetchingIsochrone: false,
      profileRequest: mockProfileRequest,
      regional: {
        bounds: mockRegion.bounds
      }
    },
    mapState: {
      activeTrips: [],
      center: {
        lat: 38.8886,
        lon: -77.043
      },
      components: [],
      zoom: 12
    },
    network: {
      error: null,
      outstandingRequests: 0
    },
    opportunityDatasets: {
      active: mockOpportunityDataset.key,
      datasets: mockRegion.opportunityDatasets,
      uploadStatuses: []
    },
    region: {
      currentRegionId: mockRegion._id,
      regions: [mockRegion]
    },
    routing: {
      locationBeforeTransitions: null
    },
    project: {
      activeModificationId: mockModification._id,
      bundles: [mockBundle],
      currentProject: mockProject,
      feeds: [mockFeed],
      feedsById: {
        [mockFeed.id]: mockFeed
      },
      modifications: [mockModification],
      projects: [mockProject],
      projectsById: {
        [mockProject._id]: mockProject
      }
    },
    user: {
      idToken: null
    }
  }
}

export const mockProjectApplicationError = {
  modificationId: 'Add Trip Pattern',
  title: "Error while applying the modification entitled 'Add Trip Pattern'.",
  messages: ['Headway is not greater than zero.']
}

export const mockRegionalAnalyses = [
  {
    ...mockModel,
    zoom: 9,
    bounds: {},
    complete: true,
    deleted: false,
    status: 'DONE',
    workerVersion: 'v3.2.0',
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 1',
    grid: 'Jobs_total',
    projectId: '1234',
    variant: 0,
    bundleId: '5678',
    cutoffMinutes: 60,
    _id: 'abcd',
    request: mockProfileRequest,
    // percentile -1 indicates this analysis was done using Andrew Owen-style average instantaneous accessibility
    travelTimePercentile: -1
  },
  {
    ...mockModel,
    zoom: 9,
    bounds: {},
    complete: true,
    deleted: false,
    status: 'DONE',
    workerVersion: 'v3.2.0',
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 2',
    grid: 'Jobs_total',
    projectId: '4321',
    variant: 1,
    bundleId: '5678',
    cutoffMinutes: 60,
    _id: 'efgh',
    request: mockProfileRequest,
    travelTimePercentile: 42
  }
]

export const mockGrid = {
  zoom: 9,
  width: 300,
  height: 300,
  north: 20024,
  west: 20015,
  data: [],
  min: 0,
  max: 2000,
  contains (x: number, y: number) { return true }
}

export const mockBookmark = {
  ...mockModel,
  _id: 'bookmarkid',
  name: 'Bookmark',
  profileRequest: mockProfileRequest,
  isochroneCutoff: 42,
  opportunityDataset: 'Workers_total',
  regionId: 'REGION'
}

export const mockSurfaceParameters = {
  width: 5,
  height: 5,
  nSamples: 5,
  zoom: 9,
  north: 40000,
  west: 40000
}

// this is a function because responseToSurface is destructive and can lead to test leakage if we
// don't have a brand-new surface each time
export const createMockSurfaceData = () => {
  const {width, height, nSamples, zoom, north, west} = mockSurfaceParameters

  const metadata = {
    projectApplicationWarnings: [mockProjectApplicationError]
  }

  const metadataEncoded = new TextEncoder().encode(JSON.stringify(metadata))
  const bufSize =
    9 * 4 + // header
    width * height * nSamples * 4 + // data
    metadataEncoded.length // metadata

  const buffer = new ArrayBuffer(bufSize)
  const surfaceData = new Int32Array(buffer, 0, 9 + width * height * nSamples)

  // text header ("ACCESSGR")
  surfaceData[0] = 1162036033 // ACCE
  surfaceData[1] = 1380406099 // SSGR

  // version
  surfaceData[2] = 0

  surfaceData[3] = zoom
  surfaceData[4] = west
  surfaceData[5] = north
  surfaceData[6] = width
  surfaceData[7] = height
  surfaceData[8] = nSamples

  // phew, header is done
  for (let pixel = 9, y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      let prev = 0
      for (let z = 0; z < nSamples; z++, pixel++) {
        const pixelVal = getMockSurfaceValue(x, y, z)
        surfaceData[pixel] = pixelVal - prev // values are delta-coded within each pixel
        prev = pixelVal
      }
    }
  }

  // copy metadata
  const metadataBuffer = new Uint8Array(
    buffer,
    buffer.byteLength - metadataEncoded.byteLength,
    metadataEncoded.length
  )
  for (let i = 0; i < metadataEncoded.length; i++) {
    metadataBuffer[i] = metadataEncoded[i]
  }

  return buffer
}

export const mockSurface = responseToSurface(createMockSurfaceData())

// Return the value that should be stored in sample z of pixel x, y of the mock surface
export function getMockSurfaceValue (x: number, y: number, z: number): number {
  // this used to simply return z, 2 * z and 3 * z when we had an array of [5, 10, 15 . . . 95]th
  // percentiles. However, when we switched to requesting specific percentiles that stopped working.
  // This translates the position in the access grid into a percentile, then subtracts five and divides
  // by five to get the position in the old array, so that tests written assuming the old structure continue
  // to pass.
  if (x < 2 && y < 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y >= 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y < 2) return 2 * (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x < 2 && y >= 2) return 3 * (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  return 0
}

const opportunityDatasetGridData = new Int32Array(
  mockSurfaceParameters.width * mockSurfaceParameters.height
)
// top half of grid is 1, bottom half is 2
for (let i = 0; i < 3 * mockSurfaceParameters.width; i++) {
  opportunityDatasetGridData[i] = 1
}
for (
  let i = 3 * mockSurfaceParameters.width;
  i < opportunityDatasetGridData.length;
  i++
) {
  opportunityDatasetGridData[i] = 2
}

export const mockDestinationGrid = {
  ...mockSurfaceParameters,
  west: mockSurfaceParameters.west - 1, // shift slightly so that we are testing misaligned grids
  north: mockSurfaceParameters.north - 1,
  data: opportunityDatasetGridData
}

export const releaseVersions = ['v3.0.0', 'v2.2.0', 'v2.0.0']
export const allVersions = [
  'v3.0.0',
  'v2.2.0-2-g789df3e',
  'v2.2.0',
  'v2.0.0-19-g589f32f',
  'v2.0.0'
]
