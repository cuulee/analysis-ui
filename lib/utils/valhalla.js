// @flow
import type {LonLat} from '../types'

import {decode as decodePolyline} from '@mapbox/polyline'
import fetch from 'isomorphic-fetch'

const MAPZEN_TURN_BY_TURN_KEY = String(process.env.MAPZEN_TURN_BY_TURN_KEY)
const COORDINATE_DIVISOR = 10

export async function route (start: LonLat, end: LonLat) {
  const json = {
    costing: 'bus',
    locations: [{lat: start.lat, lon: start.lon}, {lat: end.lat, lon: end.lon}]
  }
  const response = await fetch(
    `https://valhalla.mapzen.com/route?json=${JSON.stringify(json)}&api_key=${MAPZEN_TURN_BY_TURN_KEY}`
  )
  return response.json()
}

export async function polyline (start: LonLat, end: LonLat) {
  const json = await route(start, end)
  return decodePolyline(json.trip.legs[0].shape).map(c => [
    c[1] / COORDINATE_DIVISOR,
    c[0] / COORDINATE_DIVISOR
  ]) // Mapzen or Mapbox is encoding/decoding wrong?
}
