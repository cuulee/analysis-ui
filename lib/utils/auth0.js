// @flow
import Auth0Lock from 'auth0-lock'
import auth0 from 'auth0-js'
import {push} from 'react-router-redux'

import {setUser} from '../actions'
import messages from './messages'
import get from '../utils/get'

const CLIENT_ID = process.env.AUTH0_CLIENT_ID
const DOMAIN = process.env.AUTH0_DOMAIN
const localStorage = window.localStorage || mockLocalStorage()

export const authIsRequired = CLIENT_ID && DOMAIN

export const lock = authIsRequired
  ? new Auth0Lock(CLIENT_ID, DOMAIN, {
    auth: {
      params: {
        scope: 'openid email profile analyst offline_access'
      },
      redirect: false
    },
    allowSignUp: false,
    theme: {
      logo: 'https://s3-eu-west-1.amazonaws.com/analyst-logos/conveyal-128x128.png',
      primaryColor: '#2389c9'
    },
    closable: true,
    autoclose: true,
    languageDictionary: {
      title: messages.authentication.logIn
    }
  })
  : {
    getUserInfo () {},
    on () {},
    show () {}
  }

const client = authIsRequired
  ? new auth0.WebAuth({
    clientID: CLIENT_ID,
    domain: DOMAIN
  })
  : {}

export function bootstrap (store: any) {
  setUserAndRefreshIdToken(store)

  lock.on('authenticated', authResult => {
    if (!authResult.accessToken) {
      window.alert(`Failed to login. ${authResult.errorDescription}`)
    } else {
      lock.getUserInfo(authResult.accessToken, (error, profile) => {
        if (error) {
          store.dispatch(setUser(null))
          store.dispatch(push('/login'))
        } else {
          const user = {
            ...authResult,
            profile
          }
          localStorage.setItem('user', JSON.stringify(user))
          store.dispatch(setUser(user))
          store.dispatch(push('/'))
        }
      })
    }
  })

  store.subscribe(() => {
    const {user} = store.getState()
    if (
      (!user || !user.idToken) &&
      window.location.pathname.indexOf('/login') === -1
    ) {
      logout()
    }
  })
}

export function logout () {
  localStorage.removeItem('user') // TODO anything else in local storage?
  client.logout({
    client_id: CLIENT_ID,
    returnTo: `${window.location.origin}/login`
  })
}

function setUserAndRefreshIdToken (store) {
  const userString = localStorage.getItem('user')
  const user = userString && JSON.parse(userString)
  if (user && user.refreshToken) {
    const expDate = new Date(get(user, 'idTokenPayload.exp', user.expiresIn || 0) * 1000)
    const now = new Date()
    if (process.env.NODE_ENV !== 'development' && expDate < now) {
      logout()
    } else {
      store.dispatch(setUser(user))
    }
  } else {
    store.dispatch(push('/login'))
  }
}

function mockLocalStorage () {
  if (process.env.NODE_ENV !== 'test') {
    console.error(
      'WARNING! LocalStorage API does not exist. Mocking to prevent errors.'
    )
  }
  return {
    getItem () {},
    setItem () {},
    removeItem () {}
  }
}
