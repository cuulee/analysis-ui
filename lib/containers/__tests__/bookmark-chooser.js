// @flow
import React from 'react'

import BookmarkChooser from '../bookmark-chooser'
import {mockWithProvider} from '../../utils/mock-data'

describe('Containers > Bookmark Chooser', function () {
  it('should render correctly', function () {
    const {snapshot} = mockWithProvider(<BookmarkChooser />)
    expect(snapshot()).toMatchSnapshot()
  })
})
