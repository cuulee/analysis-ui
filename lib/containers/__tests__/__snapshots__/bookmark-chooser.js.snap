// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers > Bookmark Chooser should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(BookmarkChooser)>
    <BookmarkChooser
      bookmarkData={
        Object {
          "opportunityDataset": undefined,
          "profileRequest": Object {
            "accessModes": "WALK",
            "bikeSafe": 1,
            "bikeSlope": 1,
            "bikeSpeed": 4.166666666666667,
            "bikeTime": 1,
            "date": "2016-01-16",
            "directModes": "WALK",
            "egressModes": "WALK",
            "fromLat": undefined,
            "fromLon": undefined,
            "fromTime": 25200,
            "maxRides": 4,
            "maxTripDurationMinutes": 60,
            "monteCarloDraws": 200,
            "reachabilityThreshold": 0,
            "toTime": 32400,
            "transitModes": "BUS,TRAM,RAIL,SUBWAY",
            "travelTimePercentile": 50,
            "walkSpeed": 1.3888888888888888,
          },
          "regionId": "1",
        }
      }
      createBookmark={[Function]}
      disabled={false}
      fetchTravelTimeSurface={[Function]}
      selectBookmark={[Function]}
    >
      <Group
        label="Bookmark"
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="bookmark-input-0"
          >
            Bookmark
          </label>
          <div
            className="row"
          >
            <div
              className="col-xs-6"
            >
              <Select
                addLabelText="Add \\"{label}\\"?"
                arrowRenderer={[Function]}
                autosize={true}
                backspaceRemoves={true}
                backspaceToRemoveMessage="Press backspace to remove {label}"
                clearAllText="Clear all"
                clearRenderer={[Function]}
                clearValueText="Clear value"
                clearable={false}
                deleteRemoves={true}
                delimiter=","
                disabled={false}
                escapeClearsValue={true}
                filterOptions={[Function]}
                ignoreAccents={true}
                ignoreCase={true}
                inputProps={Object {}}
                isLoading={false}
                joinValues={false}
                labelKey="label"
                matchPos="any"
                matchProp="any"
                menuBuffer={0}
                menuRenderer={[Function]}
                multi={false}
                noResultsText="No results found"
                onBlurResetsInput={true}
                onChange={[Function]}
                onCloseResetsInput={true}
                optionComponent={[Function]}
                options={Array []}
                pageSize={5}
                placeholder="Select..."
                required={false}
                scrollMenuIntoView={true}
                searchable={true}
                simpleValue={false}
                tabSelectsValue={true}
                valueComponent={[Function]}
                valueKey="value"
              >
                <div
                  className="Select Select--single is-searchable"
                >
                  <div
                    className="Select-control"
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                  >
                    <span
                      className="Select-multi-value-wrapper"
                      id="react-select-2--value"
                    >
                      <div
                        className="Select-placeholder"
                      >
                        Select...
                      </div>
                      <mockConstructor
                        aria-activedescendant="react-select-2--value"
                        aria-expanded="false"
                        aria-haspopup="false"
                        aria-owns=""
                        className="Select-input"
                        minWidth="5"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        required={false}
                        role="combobox"
                        value=""
                      />
                    </span>
                    <span
                      className="Select-arrow-zone"
                      onMouseDown={[Function]}
                    >
                      <span
                        className="Select-arrow"
                        onMouseDown={[Function]}
                      />
                    </span>
                  </div>
                </div>
              </Select>
            </div>
            <div
              className="col-xs-6"
            >
              <Button
                block={true}
                disabled={false}
                onClick={[Function]}
                style="success"
              >
                <a
                  className="btn btn-success btn-block"
                  disabled={false}
                  href="#"
                  onClick={[Function]}
                  tabIndex={0}
                >
                  <Icon
                    type="plus"
                  >
                    <i
                      className="fa fa-plus fa-fw "
                    />
                  </Icon>
                   
                  Create bookmark
                </a>
              </Button>
            </div>
          </div>
        </div>
      </Group>
    </BookmarkChooser>
  </Connect(BookmarkChooser)>
</Provider>
`;
